dallinger/deployment.py

```diff
@@ -530,6 +526,11 @@ def deploy_sandbox_shared_setup(log, verbose=True, app=None, exp_config=None):
         "dashboard_url": "{}/dashboard/".format(heroku_app.url),
         "recruitment_msg": launch_data.get("recruitment_msg", None),
     }
+
```

Comment:
This still creates a race condition for recruitment via MTurk, etc, right? Would the alternative would be to break `/launch` into two parts, like `/launch` and `/ready` or something?

Reply:
It's a pretty small window, but it _is_ a race condition, and I don't really know with any confidence that it won't be a problem, given the automatic HIT accepting tech some (many?) Turkers use. A critical question is whether these tools are used to just _accept_ the HIT(in which case I don't the having the DB destroyed and rebuilt would matter), or if they  both accept and consent (which would definitely be very bad).

I think it would be better to have a separate route to trigger recruitment anyway. It raises more backwards compat concerns, but maybe those aren't really justified?

Reply:
If I can butt in, we've wanted for a while to have an `Experiment` property that returns whether the experiment is ready for participants yet. We often have experiments where the deployment triggers some kind of asynchronous preparation script, and we want to wait for that to complete before participants join. 

A simple implementation (that doesn't require delaying initiating recruitment) is to direct the participants to a waiting page before the consent form. On reflection, I guess a disadvantage of this though is that participants could accept the HIT long before it's ready, and then their assignment could timeout before they've had time to complete it, so maybe it's not the best plan after all... Maybe better instead to check this property every few seconds, and starting recruitment once it returns true, I don't know...

Reply:
> I've got a couple of questions. This looks pretty reasonable to me, but it does seem like there's perhaps too much indirection aimed at making some sort of generic bootstrapping mechanism, especially since it's probably being triggered too late to be generically reliable for that purpose.

I first added the archive file path as an optional argument to `deploy_sandbox_shared_setup()` and had that function perform the DB erasing and overwriting, and I hated it. It just felt overly specific in the wrong way. So, I switched to the current functional/DI thing and kind of hate that too. It's both prematurely general, and still doesn't address the problem of overwriting the DB after recruitment starts. The virtue that led me to keep it was that passing something in made testing way easier: I didn't need to add yet another monkeypatch in order to test.

Reply:
Having a testable implementation seems like a big win for this sort of thing. I like the idea of a general implementation, but I think the concept of post-launch actions is a kind of inherently misguided unless it's pre-recruitment. In this case there's a good chance it gets the job done. I guess the question is if we switch from `launch + postlaunch` to `launch + postlaunch + recruit` will we break experiments that have customized the `launch` route in unexpected ways? What other consequences might there be?

Reply:
Currently the main obstacle to being able to recruit after overwriting the DB from an archive is the MTurk recruiter's `is_in_progress` check. I'm thinking I can maybe just change the implementation of that to look for HITs with the current experiment ID in the RequesterAnnotation field, instead of local DB Participant records with "mturk" as the recruiter_id.

Reply:
I think that makes sense!

Reply:
Change to MTurkRecruiter is in place, and we're now overwriting the DB before launch/recruitment.